scio-neo4j/src/main/scala/com/spotify/scio/neo4j/syntax/SCollectionSyntax.scala

```diff
@@ -17,15 +17,84 @@
 
 package com.spotify.scio.neo4j.syntax
 
-import com.spotify.scio.neo4j.ops.{Neo4jSCollectionReadOps, Neo4jSCollectionWriteOps}
+import com.spotify.scio.coders.{Coder, CoderMaterializer}
+import com.spotify.scio.io.ClosedTap
+import com.spotify.scio.neo4j.{Neo4jIO, Neo4jOptions}
 import com.spotify.scio.values.SCollection
+import magnolify.neo4j.ValueType
+import org.apache.beam.sdk.io.{neo4j => beam}
 
-trait SCollectionSyntax {
+/** Enhanced version of [[com.spotify.scio.values.SCollection SCollection]] with Neo4J methods. */
+final class Neo4jSCollectionOps[T](private val self: SCollection[T]) extends AnyVal {
+
+  import Neo4jIO._
 
-  implicit def neo4jSCollectionWriteOps[T](sc: SCollection[T]): Neo4jSCollectionWriteOps[T] =
-    new Neo4jSCollectionWriteOps[T](sc)
+  /**
+   * Execute parallel instances of the provided Cypher query to the specified Neo4j database; one
+   * instance of the query will be executed for each element in this [[SCollection]]. Results from
+   * each query invocation will be added to the resulting [[SCollection]] as if by a `flatMap`
+   * transformation (where the Neo4j-query-execution returns an `Iterable`).
+   *
+   * This operation parameterizes each query invocation by applying a supplied function to each
+   * [[SCollection]] element before executing the Cypher query. The function must produce a [[Map]]
+   * of [[String]] to [[AnyRef]], where the keys correspond to named parameters in the provided
+   * Cypher query and the corresponding values correspond to the intended value of that parameter
+   * for a given query invocation. Named parameters must consist of letters and numbers, prepended
+   * with a `$`, as described in the Neo4j
+   * [[https://neo4j.com/docs/cypher-manual/current/syntax/parameters Cypher Manual (Syntax / Parameters)]].
+   *
+   * @see
+   *   ''Reading from Neo4j'' in the
+   *   [[https://beam.apache.org/releases/javadoc/current/org/apache/beam/sdk/io/neo4j/Neo4jIO.html Beam `Neo4jIO` documentation]]
+   * @param neo4jConf
+   *   options for configuring a Neo4J driver
+   * @param cypher
+   *   parameterized Cypher query
+   * @return
+   *   [[SCollection]] containing the union of query results from a parameterized query invocation
+   *   for each original [[SCollection]] element
+   */
+  def neo4jCypher[U](
+    neo4jConf: Neo4jOptions,
+    cypher: String
+  )(implicit
+    neo4jInType: ValueType[T],
+    neo4jOutType: ValueType[U],
+    coder: Coder[U]
+  ): SCollection[U] =
+    self.applyTransform(
+      beam.Neo4jIO
+        .readAll[T, U]()
+        .withDriverConfiguration(dataSourceConfiguration(neo4jConf.connectionOptions))
+        .withSessionConfig(neo4jConf.sessionConfig)
+        .withTransactionConfig(neo4jConf.transactionConfig)
+        .withCypher(cypher)
+        .withParametersFunction(neo4jInType.to(_).asMap())
```

Comment:
Ah nice! Yes, agree that this looks much better. 🙂

Just to confirm that my understanding is correct: it would be incumbent upon the user to write the Cypher query String such that the parameter names match the field names used by the (implicit) `ValueType[T]` implementation? (Of course, in most cases, it would probably be easiest to use a case class (with the existing `CaseMapper`-powered `ValueType`) and match the query to the case class field names.)

Reply:
Yes, that's correct.
The write API works the same way: if the cipher refers to a field which is not in the model (after the case mapper transform), there will be a runtime exception.